extends Sprite2D  

@onready var label = $Text
@onready var inv: Inv = preload("res://inventory/playerinv.tres")
@export var qid = 0 

var player_chase = false
var player = null
var is_dead = false  
var has_received_all = false  
var hold_time = 0
var cooldown_time = 0 
var received_numbers: Array = []                  
var available_numbers: Array = []
@onready var anim_sprite = $AnimatedSprite2D

func _ready(): 
	label.text = ""
	label.visible = false
	randomize()
	_set_available_numbers()

# setter ‡πÄ‡∏ß‡∏•‡∏≤‡πÄ‡∏õ‡∏•‡∏µ‡πà‡∏¢‡∏ô‡∏Ñ‡πà‡∏≤ qid
func set_qid(value):
	qid = value
	_set_available_numbers()

# ‡∏ü‡∏±‡∏á‡∏Å‡πå‡∏ä‡∏±‡∏ô‡πÑ‡∏ß‡πâ‡∏ï‡∏±‡πâ‡∏á‡∏Ñ‡πà‡∏≤‡πÄ‡∏•‡∏Ç‡∏ó‡∏µ‡πà‡∏™‡∏∏‡πà‡∏°‡πÑ‡∏î‡πâ‡∏ï‡∏≤‡∏° qid
func _set_available_numbers():
	if qid == 0:
		available_numbers = [1,2,3,4,5,10]
	elif qid == 1:
		available_numbers = [2,3,5,4,6,9]
	elif qid == 2:
		available_numbers = [5,6,7,8,9]
	# ‡πÄ‡∏û‡∏¥‡πà‡∏°‡πÄ‡∏á‡∏∑‡πà‡∏≠‡∏ô‡πÑ‡∏Ç qid ‡∏≠‡∏∑‡πà‡∏ô ‡πÜ ‡πÑ‡∏î‡πâ‡∏ï‡∏≤‡∏°‡∏ï‡πâ‡∏≠‡∏á‡∏Å‡∏≤‡∏£
	
	# ‡∏£‡∏µ‡πÄ‡∏ã‡πá‡∏ï‡∏™‡∏ñ‡∏≤‡∏ô‡∏∞
	received_numbers.clear()
	has_received_all = false
	hold_time = 0
	cooldown_time = 0

func _physics_process(delta):
	if has_received_all:
		return

	if cooldown_time > 0:
		cooldown_time -= delta
		if cooldown_time > 0:
			label.text = "‡∏£‡∏≠ " + str(int(round(cooldown_time))) + " ‡∏ß‡∏¥‡∏Å‡πà‡∏≠‡∏ô‡πÄ‡∏õ‡∏¥‡∏î‡∏Å‡∏•‡πà‡∏≠‡∏á‡πÉ‡∏´‡∏°‡πà"
			label.visible = true
		else:
			label.visible = false   # üîπ ‡∏ã‡πà‡∏≠‡∏ô‡∏Ç‡πâ‡∏≠‡∏Ñ‡∏ß‡∏≤‡∏°‡πÄ‡∏°‡∏∑‡πà‡∏≠‡πÄ‡∏´‡∏•‡∏∑‡∏≠ 0 ‡∏ß‡∏¥
		return

	
	if player_chase and is_dead:
		# ‡∏Å‡∏î F ‡∏Ñ‡πâ‡∏≤‡∏á‡πÑ‡∏ß‡πâ
		if Input.is_action_pressed("F"):
			# ‡πÉ‡∏´‡πâ‡πÄ‡∏£‡∏¥‡πà‡∏°‡πÄ‡∏•‡πà‡∏ô animation ‡πÄ‡∏â‡∏û‡∏≤‡∏∞‡∏ï‡∏≠‡∏ô‡πÄ‡∏£‡∏¥‡πà‡∏°‡∏Å‡∏î
			if not $AnimatedSprite2D.is_playing():
				$AnimatedSprite2D.play("open")

			hold_time += delta
			label.text = "‡∏Å‡∏≥‡∏•‡∏±‡∏á‡πÄ‡∏õ‡∏¥‡∏î‡∏Å‡∏•‡πà‡∏≠‡∏á" +"\n"+  str(int(round(hold_time))) + " / 5 ‡∏ß‡∏¥‡∏ô‡∏≤‡∏ó‡∏µ"
			label.visible = true
			$Text2.text = "" 
			
			if hold_time >= 2:
				if available_numbers.size() > 0:
					var random_num = available_numbers.pick_random()
					available_numbers.erase(random_num)
					received_numbers.append(random_num)
					
					print("‡∏™‡∏∏‡πà‡∏°‡πÑ‡∏î‡πâ‡πÄ‡∏•‡∏Ç: ", random_num)
					print("‡πÄ‡∏•‡∏Ç‡∏ó‡∏µ‡πà‡πÑ‡∏î‡πâ‡∏£‡∏±‡∏ö‡πÅ‡∏•‡πâ‡∏ß: ", received_numbers)
					print("‡πÄ‡∏•‡∏Ç‡∏ó‡∏µ‡πà‡πÄ‡∏´‡∏•‡∏∑‡∏≠‡πÉ‡∏´‡πâ‡∏™‡∏∏‡πà‡∏°: ", available_numbers)
					$"../player/Inv_UI2".add_choice_to_inventory(random_num)
					$Text2.text = "‡∏Ñ‡∏∏‡∏ì‡πÑ‡∏î‡πâ‡∏£‡∏±‡∏ö‡∏õ‡∏∏‡πà‡∏°‡∏´‡∏°‡∏≤‡∏¢‡πÄ‡∏•‡∏Ç " + str(random_num)
				else:
					has_received_all = true
					label.text = "‡∏Ñ‡∏∏‡∏ì‡πÑ‡∏î‡πâ‡∏£‡∏±‡∏ö‡∏ï‡∏±‡∏ß‡πÄ‡∏•‡∏Ç‡∏ó‡∏±‡πâ‡∏á‡∏´‡∏°‡∏î‡πÅ‡∏•‡πâ‡∏ß!"
				hold_time = 0.0
				cooldown_time = 2
				$collectnumber.play()

		# ‡∏õ‡∏•‡πà‡∏≠‡∏¢‡∏õ‡∏∏‡πà‡∏° F
		elif Input.is_action_just_released("F"):
			$AnimatedSprite2D.stop()
			$AnimatedSprite2D.frame = 0  # ‡∏Å‡∏•‡∏±‡∏ö‡πÑ‡∏õ‡πÄ‡∏ü‡∏£‡∏°‡πÅ‡∏£‡∏Å (‡πÄ‡∏•‡∏∑‡∏≠‡∏Å‡πÉ‡∏™‡πà‡∏´‡∏£‡∏∑‡∏≠‡πÑ‡∏°‡πà‡∏Å‡πá‡πÑ‡∏î‡πâ)
			hold_time = 0.0


func _on_detection_area_body_entered(body):
	if body.name == "player":
		player_chase = true
		player = body
	
		if not has_received_all:
			label.text = "‡∏Å‡∏î F ‡∏Ñ‡πâ‡∏≤‡∏á‡πÄ‡∏û‡∏∑‡πà‡∏≠‡πÄ‡∏õ‡∏¥‡∏î‡∏Å‡∏•‡πà‡∏≠‡∏á"
		else:
			label.text = "‡∏Ñ‡∏∏‡∏ì‡πÑ‡∏î‡πâ‡∏£‡∏±‡∏ö‡∏ï‡∏±‡∏ß‡πÄ‡∏•‡∏Ç‡∏Ñ‡∏£‡∏ö‡πÅ‡∏•‡πâ‡∏ß!"
		label.visible = true
		is_dead = true

func _on_detection_area_body_exited(body):
	if body.name == "player":
		player_chase = false
		label.visible = false
		hold_time = 0
